rust/main/chains/hyperlane-ethereum/src/rpc_clients/fallback/tests.rs

```diff
@@ -0,0 +1,216 @@
+
+    fn deref(&self) -> &Self::Target {
+        &self.provider
+    }
+}
+
+impl EthereumProviderMock {
+    fn new(request_sleep: Option<Duration>, success: bool) -> Self {
+        Self {
+            provider: ProviderMock::new(request_sleep),
+            success,
+        }
+    }
+}
+
+impl From<EthereumProviderMock> for JsonRpcBlockGetter<EthereumProviderMock> {
+    fn from(val: EthereumProviderMock) -> Self {
+        JsonRpcBlockGetter::new(val)
+    }
+}
+
+fn dummy_success_return_value<R: DeserializeOwned>() -> Result<R, HttpClientError> {
+    serde_json::from_str("0").map_err(|e| HttpClientError::SerdeJson {
+        err: e,
+        text: "".to_owned(),
+    })
+}
+
+fn dummy_error_return_value<R: DeserializeOwned>() -> Result<R, HttpClientError> {
+    serde_json::from_str("not-a-json").map_err(|e| HttpClientError::SerdeJson {
+        err: e,
+        text: "".to_owned(),
+    })
+}
+
+#[async_trait]
+impl JsonRpcClient for EthereumProviderMock {
+    type Error = HttpClientError;
+
+    /// Pushes the `(method, params)` to the back of the `requests` queue,
+    /// pops the responses from the back of the `responses` queue
+    async fn request<T: Debug + Serialize + Send + Sync, R: DeserializeOwned>(
+        &self,
+        method: &str,
+        params: T,
+    ) -> Result<R, Self::Error> {
+        if let Some(sleep_duration) = self.provider.request_sleep() {
+            sleep(sleep_duration).await;
+        }
+        self.push(method, params);
+        if self.success {
+            dummy_success_return_value()
+        } else {
+            dummy_error_return_value()
+        }
+    }
+}
+
+impl PrometheusJsonRpcClientConfigExt for EthereumProviderMock {
+    fn node_host(&self) -> &str {
+        todo!()
+    }
+
+    fn chain_name(&self) -> &str {
+        todo!()
+    }
+}
+
+impl<C> EthereumFallbackProvider<C, JsonRpcBlockGetter<C>>
+where
+    C: JsonRpcClient<Error = HttpClientError>
+        + PrometheusJsonRpcClientConfigExt
+        + Into<JsonRpcBlockGetter<C>>
+        + Clone,
+    JsonRpcBlockGetter<C>: BlockNumberGetter,
+{
+    async fn fallback_test_call(&self) {
+        self.request::<_, u64>(BLOCK_NUMBER_RPC, ()).await.unwrap();
+    }
+}
+
+#[tokio::test]
+async fn test_multicast_first_provider_succeeds() {
+    let fallback_provider_builder = FallbackProviderBuilder::default();
+    let providers = vec![
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+        EthereumProviderMock::default(),
+    ];
+    let fallback_provider = fallback_provider_builder.add_providers(providers).build();
+    let ethereum_fallback_provider = EthereumFallbackProvider::new(fallback_provider);
+    ethereum_fallback_provider
+        .multicast::<_, u64>(BLOCK_NUMBER_RPC, ())
+        .await
+        .unwrap();
+    let provider_call_count: Vec<_> =
+        ProviderMock::get_call_counts(&ethereum_fallback_provider).await;
+    assert_eq!(provider_call_count, vec![1, 0, 0]);
```

Comment:
why would we expect the first provider to succeed and not the others?

confused for a couple reasons:
- wouldn't we expect all 3 to be queried concurrently? I guess only the first one is queried because with this mock setup we never end up with a future that isn't ready so it just immediately resolves to the first completed 0th provider future before trying any of the others?
- I would've thought the use of `FuturesUnordered` wouldn't guarantee that the first one always wins the race?

Reply:
makes sense! might be nice to have a quick comment for future readers but happy to merge without it